src/effects/PerTrackEffect.cpp

```diff
@@ -117,10 +128,23 @@ void PerTrackEffect::Buffers::Discard(size_t drop, size_t keep)
       position = *++iterP;
       memmove(position, position + drop, size);
    }
+#else
+   // Version that assumes the precondition
+   assert(drop + keep <= Remaining());
```

Comment:
Please read this before I make any change https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1743r0.pdf

There SHOULD be UB when preconditions fail.  That is the point of making them explicit and making clear which code is at fault when a bug happens.

The bounds checks were only lately added by me in code that never had a bug in them when I didn't yet understand why.  Now I have added more argument, and checkable assertions in debug builds, and I'm only returning it to being unchecked as it was.  I did the work of making more explicit a lot of implicit, unexplained assumptions that made the function work.

Having made it explicit, I can also now recheck my assumptions more carefully as I make other transformations.

Reply:
If you reviewed this, I expect you to point to an error in my proofs or tell me you have reasoned through them and understood.

Reply:
I do not understand how a proposal not yet accepted is a valid reason behind removing run-time checks. Audacity is a production software used in real world, UB in processing code is exceptionally bad idea. Much worse, than memory allocation in RT thread, so to speak.

>The bounds checks were only lately added by me in code that never had a bug in them when I didn't yet understand why.

This is very hard to understand that from the PR.

>If you reviewed this

I do not consider a code review to be a proof good enough.  Lately, the master was broken far too often even after the heavy reviews, including debugging sessions, which are not in scope of the code review and which slow down the development considerably.  You ask me to formally validate a non trivial change inside a large class used inside a huge code base, but I do not trust it to be possible at all and for sure - I, personally, will miss a lot even if I will study all the scenarios that currently can happen. The same way as I approved the code that broke generators, because at that point not you nor I though of the case. So probably you need a different reviewer for this PR.

I do not agree that it is possibly to formally prove anything in a huge, old code base with a zero unit tests coverage.

Reply:
By the way, I do not agree with the examples that paper uses either. Crash is not an UB. It might be undesirable, but it is very well defined. Their first example does have a UB which is not feasible to check though, when str points to a random region of memory. In this case - we guarantee by construction that the memory region is valid, but we potentially allow overwriting past the buffer, which is much worse. Please note, I do not say we DO overwrite, but I cannot verify (probably some else can) that we do not.

Reply:
Each time I add an assertion within a function, I mean it to be provable either from preconditions or from deduction about the effects of preceding lines, and I add comments why those assertions are justified.  I'm just expecting a reviewer to agree that each one of these deductions is clear enough.

The added assertions improve old code like this that left too many important assumptions unstated, and help prevent fragility as I or someone else apply more tranformations to the code.  As indeed I will in commits soon to come.

I am also trying to raise confidence in the correctness of those commits to come and I don't think I'm wasting time by adding many assertions.

Reply:
>I do strongly recommend the linked paper.

I have read it multiple times. I do not agree with it fully and I consider their examples to be not representative. P1744 is to some extent implemented in MS STL and causes heavy performance issues in debug builds while giving a sense of false safety in release. I sincerely hope that P1743 and P1744 will be revised heavily before being accepted into the C++ standard if accepted at all.

Reply:
All right, we don't agree on all, but you have read it.

The paper includes one good thing we should agree on:  it explains when NOT to use assertions.  They are wrong for validating input from the outside world.  Some old Audacity code mis-used assertions that way.  See 735c4faa34790b6c4ae80e7743566b74c7fad8bc and 5e21f526b49edcbfa8af8d36caef94b366d35d73 and 046ae403ba9b2d20889eb09cfa884aa324575d45

Please note the bounds checks I only lately added were an effort to add defensive code in cases where I didn't yet have the proof of preconditions I was seeking.  The code was in fact all right without them, relying on unstated assumptions.  I can leave it as it was, or keep the bounds checks, against the possibility that the Buffers classes become more widely used, but which they are not, yet.

Reply:
Please understand and do agree:  when I write a new assertion, I mean it to be provable, and I want my reviewer to challenge my assertions if they are not accompanied by sufficient comments to make it obvious from context why they are right.

Reply:
Also if your reconsider commit 105929e728438cc181ad55edc107cc031932e068 which fixed the generator problem:  I found that I needed to guard agains UB with a test on `numAudioIn` in case of processors by exiting early because there was no proof of the correctness of a value returned by a virtual function (the set of overrides is open-ended, NOT what you can review in one self-contained file.)

I was adding a defense against possible UBs, which you should like.  The fault was defending in some unnecessary cases too and causing error returns (not more UBs).